ntp-daemon/src/main.rs

```diff
@@ -1,169 +1,165 @@
-                                system_snapshot.leap_indicator = clock_select.system_peer_snapshot.leap_indicator;
-                                system_tx.send(system_snapshot).expect("System snapshot mechanism failed");
-                            }
-                        }
-                    }
-                    None => info!("filter and combine did not produce a result"),
+        // remove snapshots from previous iteration
+        snapshots.clear();
+
+        // add all valid measurements to our list of snapshots
+        snapshots.extend(peers.iter().filter_map(PeerState::get_snapshot));
+
+        let ntp_instant = NtpInstant::now();
+        let system_poll = global_system_snapshot.read().await.poll_interval;
+        let result = FilterAndCombine::run(&config, &snapshots, ntp_instant, system_poll);
+
+        let clock_select = match result {
+            Some(clock_select) => clock_select,
+            None => {
+                info!("filter and combine did not produce a result");
+                continue;
+            }
+        };
+
+        let offset_ms = clock_select.system_offset.to_seconds() * 1000.0;
+        let jitter_ms = clock_select.system_jitter.to_seconds() * 1000.0;
+        info!(offset_ms, jitter_ms, "system offset and jitter");
+
+        let adjust_type = controller.update(
+            clock_select.system_offset,
+            clock_select.system_jitter,
+            clock_select.system_peer_snapshot.root_delay,
+            clock_select.system_peer_snapshot.root_dispersion,
+            clock_select.system_peer_snapshot.leap_indicator,
+            clock_select.system_peer_snapshot.time,
+        );
+
+        // Handle situations needing extra processing
+        match adjust_type {
+            ClockUpdateResult::Panic => {
+                panic!(
+                    r"Unusually large clock step suggested,
+                            please manually verify system clock and reference clock 
+                                 state and restart if appropriate."
+                )
+            }
+            ClockUpdateResult::Step => {
+                for peer_state in peers.iter_mut() {
+                    peer_state.reset();
                 }
+
+                reset_epoch = reset_epoch.inc();
+                reset_tx.send_replace(reset_epoch);
+            }
+            _ => {}
+        }
+
+        // Handle updating system snapshot
+        match adjust_type {
+            ClockUpdateResult::Ignore => {}
+            _ => {
+                let mut global = global_system_snapshot.write().await;
+                global.poll_interval = controller.preferred_poll_interval();
+                global.leap_indicator = clock_select.system_peer_snapshot.leap_indicator;
+            }
+        }
+    }
+
+    // the channel closed and has no more messages in it
+    Ok(())
+}
+
+fn receive_msg_for_system(
```

Comment:
I think the `MsgForSystem` name is a little unclear, with the usage of `Msg` being unidiomatic (Rust type names usually avoid abbreviations). I could probably call it `PeerUpdate`. I would also suggest making the outer message type a named field struct so that `sender` (to replace `index`) can be a field you can use without matching first. Could potentially be a `struct PeerUpdate<T> { sender: usize, message: T }` as well if there are multiple kinds of channels (but that doesn't seem to apply here/yet).

Personally I prefer inlining relatively short single-use functions like this, since I find code easier to follow if I don't have to jump around. On the other hand, if you end up defining a `Peers` type as suggested above, this could plausibly be a method on that.

Reply:
> with the usage of Msg being unidiomatic 

this is just not true. See `Vec`, `RwLock`, `Cow`. I generally like writing out names but especially in type names abbreviations are commonly used in practice.

Anyhow, the problem I'm hitting here is that most names don't communicate direction: does the system want the peer to update, or has the peer updated, which should cause changes to the system. "ForSystem" makes clear who needs to deal with this message.

Reply:
I commonly use acronyms in type names, but not abbreviations. `Vec`, in my mind, is an exception because it's extremely common. I read `PeerUpdate` as a "peer update", which I think has a clear direction. That said, I'm not that interested in bikeshedding, so feel free to use my suggestions or ignore them.